x-pack/plugins/enterprise_search/public/applications/app_search/components/engines/engines_logic.test.ts

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { LogicMounter } from '../../../__mocks__/kea.mock';
+
+jest.mock('../../../shared/http', () => ({
+  HttpLogic: { values: { http: { get: jest.fn() } } },
+}));
+import { HttpLogic } from '../../../shared/http';
+
+import { EngineDetails } from '../engine/types';
+import { EnginesLogic } from './';
+
+describe('EnginesLogic', () => {
+  const DEFAULT_VALUES = {
+    dataLoading: true,
+    engines: [],
+    enginesTotal: 0,
+    enginesPage: 1,
+    metaEngines: [],
+    metaEnginesTotal: 0,
+    metaEnginesPage: 1,
+  };
+
+  const MOCK_ENGINE = {
+    name: 'hello-world',
+    created_at: 'Fri, 1 Jan 1970 12:00:00 +0000',
+    document_count: 50,
+    field_count: 10,
+  } as EngineDetails;
+  const MOCK_ENGINES_API_RESPONSE = {
+    results: [MOCK_ENGINE],
+    meta: {
+      page: {
+        current: 1,
+        total_pages: 10,
+        total_results: 100,
+        size: 10,
+      },
+    },
+  };
+
+  const { mount } = new LogicMounter(EnginesLogic);
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('has expected default values', () => {
+    mount();
+    expect(EnginesLogic.values).toEqual(DEFAULT_VALUES);
+  });
+
+  describe('actions', () => {
+    describe('onEnginesFetch', () => {
+      describe('dataLoading', () => {
+        it('should be set to false', () => {
```

Comment:
@byronhulcher - this is how we've been writing our Kea logic file tests. Jason set up this architecture/pattern all the way back when the [credentials logic](https://github.com/elastic/kibana/blob/master/x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts) was migrated, and it's been serving us well so far. It's extremely explicit but as a result it's easy to read/scan and write. Here's an example pattern (with exx:

```js
  describe('actions', () => {
    // We don't have a separate block for 'listeners' but recently I've been debating that we should

    describe('someAction', () => {
      describe('aValueTheActionAffects', () => {
        it('should be set to the provided value', () => { 
          // mount the logic file, run the action, assert the value
        });
      });
      describe('anotherValueTheActionAffects', () => {
        it('should be set to the provided value', () => { 
          // mount the logic file, run the action, assert the value
        });
      });
    });

    describe('someOtherAction', () => {
      describe('aValueTheActionAffects', () => {
        it('should be set to the provided value', () => { 
          // mount the logic file, run the action, assert the value
        });
      });
    });

  });

  describe('selectors', () => {
    // We do have a separate section for selectors if we're testing those

    describe('someSelector', () => {
      describe('the selector changes to X when value does Y', () => {
        // ...
      });
    });

  });
```

LMK if that makes sense or feel free to comment/ask questions!

Reply:
I also think it's served us fairly well. There's still some room for improvement. I think in the future, it'd be nice if we keep the action blocks as we do now, but try to consolidate the number of reducer blocks.

Each action test would start with the following reducer block:

```
    describe('onEnginesFetch', () => {
        it('should set state', () => {
          mount();
          EnginesLogic.actions.onEnginesFetch({ engines: [], total: 0 });

          expect(EnginesLogic.values).toEqual({
            engines: [],
            dataLoading: false,
          });
    }
```

If we have reducers with multiple code paths, then we can spin up additional reducer blocks:

```
    describe('onEnginesFetch', () => {
        it('should set state', () => {
          mount();
          EnginesLogic.actions.onEnginesFetch({ engines: [], total: 0 });

          expect(EnginesLogic.values).toEqual({
            engines: [],
            dataLoading: false,
          });
        }

        describe('engines', () => {
          it('won't be set if xyz is true ', () => {
            mount();
            EnginesLogic.actions.onEnginesFetch({ engines: [], total: 0, xyz: true });

            expect(EnginesLogic.values).toEqual({
              ...DEFAULT_VALUES,
              engines: undefined,
            });
        }
    }
```

Just thinking out loud here.

Reply:
I like that a lot! Huge +1